dawn/src/dawn/Optimizer/PassSetStageLocationType.h

```diff
@@ -0,0 +1,47 @@
+//===--------------------------------------------------------------------------------*- C++ -*-===//
+//                          _
+//                         | |
+//                       __| | __ ___      ___ ___
+//                      / _` |/ _` \ \ /\ / / '_  |
+//                     | (_| | (_| |\ V  V /| | | |
+//                      \__,_|\__,_| \_/\_/ |_| |_| - Compiler Toolchain
+//
+//
+//  This file is distributed under the MIT License (MIT).
+//  See LICENSE.txt for details.
+//
+//===------------------------------------------------------------------------------------------===//
+
+#pragma once
+
+#include "Pass.h"
+
+namespace dawn {
+
+/// @brief
+/// * Input: any IIR where GridType == Unstructured and where statements inside the same stage are
+/// on the same location type (local variables must have a location type and no scalar variables are
+/// allowed).
+/// * Output: same IIR, but each stage will have it's location type correctly set (!= std::nullopt)
+/// to the location type of its statements
+/// @ingroup optimizer
+///
+/// This pass is necessary to generate legal IIR for GridType == Unstructured.
```

Comment:
can this not also set it to cells trivially for structured grids?

Then this pass could run for both and we would not have an additional branch of flow?

Reply:
It could, however I'd prefer clarity over not-branching. Running a path which does essentially nothing seems confusing to me.

Reply:
The location type of a stage is deduced from the lhs of the first (useful) statement. Querying the location type of a cartesian field doesn't make sense and it's also not allowed with our current data structures (`sir::FieldDimensions`).
Changing the deduction to consider cartesian as special case is again branching (hidden inside the pass). Changing the data structures to fit this case seems a bad option.

Reply:
The advantage of running it also for structured would be that you have a clear contract on how the IIR would look after the pass (that every stage has a location type. Then we could consider checking for this property after that (maybe after the parallel group) which would not work when we're not doing it for structured grids.

Reply:
and I don't understand why cartesian does not just have cells as the location type? Then deduction works the same way.
Can you elaborate on what would need to change?

Reply:
LocationTypes seem like a fundamental feature that should be in both structured and unstructured grids. I was concerned from the get-go when this wasn’t in the structured grids. As you say it could just be set to cells.

Reply:
@twicki we already discussed it (also with @cosunae ) when we were working on https://github.com/MeteoSwiss-APN/dawn/commit/ccfc9818965cdbe540fd4a092a69089e10ecb8db and we needed to refactor the field dimensions. The decision was to split the horizontal part in cartesian (IJ masks) and unstructured (dense location type + sparse). 
To change it back, we would need to change the protobuf (and consequently all serialized sir/iir files), the field dimensions classes and their APIs (and their usages). Also, local variable types (OnCells, OnEdges, ..., OnIJ) were modeled following this decision and would imply more changes.

Not taking into consideration the amount of work needed to change, my opinion is that we shouldn't share a concept which is not needed for one of the gridtypes. 
Also, I'd say having few gridtype-specific passes might be inevitable (I'm thinking about, for example, optimizing away a sparse field by inlining it into a reduction where it's used), while still remains that the vast majority of passes are shared.
I agree that branching is not ideal, but I'm confident that, once we'll have enough integration/end-to-end test also for unstructured, it won't cause us problems.

Reply:
I see, thanks for the explanation